File: django_app/iss/iss/surveys/checkboxWidget.py
from iss.surveys.abstractWidgetSubclass import AbstractWidgetSubclass


class CheckboxWidget(AbstractWidgetSubclass):
    args_data = {
        'name': {
            'type': 'string',
            'required': 1
        },
        'data': {
            'type': 'listWithTuples',
            'required': 1,
            'args': {
                    'type': 'tupleWithoutLists',
                    'args': {
                        'id': {
                            'type': 'string',
                            'required': 1
                        },
                        'text': {
                            'type': 'extendedString',
                            'required': 1
                        },
                        'checked': {
                            'type': 'bool',
                            'required': 0,
                            'default': 'false'
                        }
                    },
                'unnamedArgs': ['id', 'text', 'checked'],
            }
        },
        'horizontal': {
            'type': 'bool',
            'required': 0,
            'default': 'false'
        },
        'required': {
            'type': 'bool',
            'required': 0,
            'default': 'false'
        },
    }

    unnamed_args = ['name', 'data', 'horizontal', 'required']

File: django_app/iss/iss/static/surveys/js/checkboxWidget.js
$.widget("iss.checkboxWidget", $.iss.widget, {
    options: {
        questionWidget: null,
        condition: null,
        name: function() { return "" }, 
        data: function() { return [] },
        horizontal: function() {return false },
        required: function() {return false },
        resultVarName: null,
        tupleArgs: {
            id: function() { return "" }, 
            text: function() { return "" }, 
            checked: function() { return false }
        }
    },

    _create: function() {
        this._super();
    },

    _setOption: function(key, value) {
        if (key == "data")
            this._setData(value);
        if (key == "resultVarName" && value != null)
            this._setCallback(value);
        this._super(key, value);
    },

    _setData: function(dataFun) {
        this.element.find("label").remove();
        var data = dataFun();
        for (var i = 0; i < data.length; i++) {
            var checkbox = data[i]();
            this.updateTupleArgs(checkbox, this.options.tupleArgs);
            $('<label ' + (this.options.horizontal() ? 
                    '' : 'style="display:block;"') + '>'
                + '<input type="checkbox" name="'
                + this.options.name() + '" id="'
                + checkbox['id']() + '" '
                + (checkbox['checked']() ? 'checked' : '')
                + '>' + checkbox['text']() + '</label>')
                .appendTo(this.element);
        }
    },

    _setCallback: function(varName) {
        var element = this.element;
        function getChecked() {
            var checked = [];
            element.find("input:checked").each(function () { 
                checked.push($(this).attr("id")); 
            });
            return checked;
        }
        iss.vars[varName] = getChecked();
        var name = this.options.name();
        var questionWidget = this.options.questionWidget;
        $('input[name=' + name + ']').click(
            function() {
                iss.vars[varName] = getChecked();
                questionWidget.childChanged();
            });
    },
    
    validate: function() {
        if (this.options.required()
            && this.containsSubmitData
            && this.element.find("input:checked").length == 0) {
            this.element.addClass('erorr');
            return false;
        }
        return true;
    },

    insertSubmitData: function(submitData) {
        if(this.containsSubmitData) {
            var checked = [];
            this.element.find("input:checked").each(function () {
                checked.push($(this).attr("id"));
            });
            submitData[this.options.name()] = checked;
        }
    }
    
});

